### DISCUSSION TITLE
Roku power button

### DISCUSSION THREAD

**Discussion by Wolfang42** [PRProd/HA-Firemote#429] [Q&A] [answered]
I love this integration! Thank you for your work!

Is it possible and if so, how, to program the power button on the Roku remote image to power toggle a TV in the same room (in this case a Panasonic Viera GT30 that I cannot find a remote integration for) ??

**Comment by PRProd** [comment]
Greetings @Wolfang42!

Thank you for your kind words!  I really appreciate them!  The answer to your question is yes!  Let me take you through some options:

### 1. Native Home Assistant Control
You stated that you could not find an integration for your Panasonic Viera GT40, but if you do run across something in HACS or a new HA release that can control your TV, this is a good option.  Because of the way [Firemote sends commands to your devices](https://github.com/PRProd/HA-Firemote/wiki#how-does-firemote-send-commands-to-my-devices), it is always reliant on integrations.  If you find an integration that can toggle the power of your TV, then it would be easy to use a Firemote [button override](https://github.com/PRProd/HA-Firemote?tab=readme-ov-file#button-overrides) that can do what you are asking.
<br>

### 2. CEC
If your Panasonic Viera GT30 is connected directly to your Roku via a HDMI cable, there is a strong chance that you can enable some options to make this work.  Options might need to be enabled on both your TV and your Roku device.

From [this article](https://support.roku.com/article/360034303013) I found some instructions that will likely apply to your TV:

![image](https://github.com/PRProd/HA-Firemote/assets/25486599/79d4e86f-e2da-469c-a3c4-edcf5685e4e5)

Depending on the type of Roku device you have, the "1-touch play" or "CEC" settings may be in a different place.  On my Roku device, the setting you will need to enable is in **Settings** -> **System** -> **Control other devices** -> **1-touch play**

![Screenshot 2024-06-01 17-58-07](https://github.com/PRProd/HA-Firemote/assets/25486599/6c78de27-5b72-41c3-874b-b0d77295951a)

If you have CEC enabled properly, you should be able to turn your TV on and off with your Roku's physical remote control.  Once you are able to do that, Firemote will automatically do it too!
<br>

### 3. Broadlink or other IR blaster device
If the only way you can find to control your TV is with its regular IR remote control, you might consider buying an IR blaster such as a Broadlink device that can integrate with Home Assistant, and is capable of learning and repeating IR remote commands.  I must warn you, it's not quick or easy to configure a broadlink device with Home Assistant, so keep that in mind if you choose this route.  Once you are able to control your TV through the Home Assistant broadlink integration and device, then the next step is to use a Firemote [button override](https://github.com/PRProd/HA-Firemote?tab=readme-ov-file#button-overrides) to map the Firemote button to the proper broadlink command.

I purchased this one recently.  It works, but it's not easy to work with.  https://a.co/d/cvcbSTf
<br>

### 4. Brute Force
This might seem like a silly idea, but the cost is a lot lower than a Broadlink device.  Does your TV turn on as soon as the power cord is plugged in?  Is there a menu option on your TV that would enable this behavior?  If so, a simple wi-fi outlet switch like [this one](https://a.co/d/44BhTfw) actually integrates well with Home Assistant.  I have several of these in my house.  Again, like with options 1 and 3, once you get this device to properly toggle the power on and off for your TV through Home Assistant, a simple Firemote [button override](https://github.com/PRProd/HA-Firemote?tab=readme-ov-file#button-overrides) would be the final step.
<br>

Hopefully one of these solutions works for you.  Please let me know!  I'm excited to hear back from you.

-Doug

**Reply by Wolfang42** [reply]
Thank you so much for the reply! These are very informative and I still have much to learn.

The Panasonic integration does allow me to turn off the tv but not turn it on. Unfortunately, the Roku is distributed to the TV via a Savant matrix. I am trying to extricate myself from Savant entirely using Home Assistant.  Currently, a Savant scene triggered via  Alexa turns on the TV and the LG soundbar and the Roku. I assume it is HDMI-Cat6 adapter driven routed through a switch.  But I was hoping since the Panasonic button in HA can turn off the TV, there would be a similar way to turn it on.

**Comment by PRProd** [comment]
@Wolfang42 ~

Oh!  You can turn it off through HA?  This could be important!

I've heard of this behavior for TVs or media players where HA can turn it off but not turn it on again.  Sometimes this behavior can be changed in the device's power or network option menus.  Some devices have an option of "eco mode" where powering off the device will completely power it down.  Some devices also have an option for more of a "standby" or "quick start" mode where less power is consumed, the screen is off (so the device _appears_ to be off), but the network jack, IR port, and/or the CEC commands from an HDMI cable are still being listened to.

When the device is truly "off", the network jack / wifi connections are disabled, so you aren't able to send any commands to them.  However, if you have the option to make your device keep those connections alive when the screen is off (standby/ quick start mode) that could solve everything!

I found a users manual for a Panasonic Viera GT30 [here](https://www.manualslib.com/manual/873688/Panasonic-Viera-Tc-P50gt30.html#product-VIERA%20TC-P55GT30) and I quickly searched through it to see if I could find the correct setting.  I found this on page 53:

![image](https://github.com/PRProd/HA-Firemote/assets/25486599/ab951bf7-8513-440a-9947-a16c84f09251)

Which lead me to this diagram on page 34:

![viera_tcp50gt30](https://github.com/PRProd/HA-Firemote/assets/25486599/0862af96-184e-420c-8fa7-38b0f28b3f44)
<br>

I have a **STRONG** feeling that if your TV isn't currently set to **Quick Start** right now, and if you changed it to **Quick Start** you might have success with turning your TV off AND on with Home Assistant!  

Please let me know!!

**Reply by Wolfang42** [reply]
You are so thorough!  Brilliant!

But I had already done those things and double checked in the settings. When I tried today, I could not even turn it off with HA. I then realized that it was not set to a static IP and so HA could not find it anymore. Once I set it to a static IP and reinstalled  the integration, it worked fine turning off. But as soon as I turn it off in HA, it goes unavailable. That off signal truly powers it all the way off and off the network. Perhaps this model TV doesn't have a standby mode? 

Savant is controlling it through a PureLink Cat6-HDMI adapter/controller to allow different video distributed sources to go to this endpoint. Perhaps that is blocking. Can't wait to eventually be rid of Savant entirely.

**Reply by PRProd** [reply]
@Wolfang42 ~

Bummer!  :sob:  That is so disappointing!!  But there's hope!  I was thinking about this, and I wondered if using a "wake on lan" packet through Home Assistant would turn on your TV.  I was going to send you a link to the [Wake on LAN](https://www.home-assistant.io/integrations/wake_on_lan/) documentation and suggest that you try it out... but then I thought to myself, I wonder if whatever Panasonic integration is being used has some insight about this.  AND I WAS RIGHT!  ON BOTH COUNTS!

Have you seen this?  https://www.home-assistant.io/integrations/panasonic_viera/  

Maybe the Panasonic Vera integration is the one that you are using.  If so,the documentation that's provided has some insight that might help:

![image](https://github.com/PRProd/HA-Firemote/assets/25486599/019841ff-015d-4b33-8157-32d879edef2a)

![image](https://github.com/PRProd/HA-Firemote/assets/25486599/ffd2dc68-e0fd-46ed-9ded-65a0a0bb9484)

I have renewed hope now!  :smile:  I'm **very** interested to know if this works for you.  I feel invested in your success now.  :smile: 

-Doug

**Comment by Wolfang42** [comment]
I love your enthusiasm and optimism!  I might just catch it!

But I'm a pessimist my nature. I read both Wake on Lan and the Panasonic integration sites. I've assigned a static IP to the address of the TV, but the TV lists a different MAC than the router does for that IP. Perhaps the MAC is of the PureLink the TV is attached to instead?

I've created the Wake button and assigned it to the IP and MAC and just in case duplicated it to the MAC that the TV lists as well as the MAC the router lists at that IP.  So far no response.  Once I turn it off in the Panasonic Viera integration, it is unavailable until I turn it on with the hardware remote or a Savant scene to trigger it on.

There may be a way to make this work with the current hardware but at this point it is beyond me.  All my HA work so far is with integrations not requiring YAML coding or via NodeRed.  

I appreciate your patience with me!

**Reply by PRProd** [reply]
So does that mean that your TV's ethernet cable is plugged into the savant unit?  If so, does it need to be?  If you are able to run the cable directly to your LAN/switch/router outside of the savant ecosystem, maybe that could change things?

**Reply by Wolfang42** [reply]
The Savant controller, the router and switch are all upstairs. The TV ethernet cable is plugged into the PureLink device into which the Cat6 from Savant via a switch is also plugged .

**Reply by PRProd** [reply]
Man... another blocker.  

I'm starting to believe now that an IR blaster is the best route to making this work without removing your Savant from the equation.  UNLESS that cheesy power switch idea I had would actually work for you, and you'd be OK with it.  They aren't great options, but one or both of them might be the only way to make this work.

**Reply by Wolfang42** [reply]
Well I definitely want to get rid of Savant...that is the end game...then all is easy. I'll save this for last and reconnect with you when at that point. Have to redo the network to get to that point. I appreciate you!

**Reply by PRProd** [reply]
Thanks @Wolfang42!  I appreciate you and your conversation too!

I think we may have overlooked one important thing here: Savant can turn your TV on using an Alexa routine.  Have you tried triggering an Alexa routine through Home Assistant?  I assume it's possible, but I haven't had the need to research it for myself so I don't know.

https://github.com/alandtse/alexa_media_player/wiki#run-custom-command

**Reply by Wolfang42** [reply]
Yes I control Savant scenes now via Alexa. I had not thought of creating a button on the dashboard to activate a scene!  That's a genius idea in the interim while I await the great nuking of Savant.  Now how to assign that to the on/off button on the  Roku remote?

**Reply by PRProd** [reply]
That part is easy!  :smile:  Use a [button override](https://github.com/PRProd/HA-Firemote?tab=readme-ov-file#button-overrides)

The way I would go about it at first is to write a Home Assistant script that toggles the power of your Roku via the Alexa.  I would start here for safety reasons.  If the script works, then you know you can add the Firemote part.  If the script doesn't work, then the Firemote part wouldn't work either.

Once the script is working, all you need to do is add the script as a button override to your Firemote YAML config, like this:

```yaml
button_overrides:
  power-button:
    script: { THE NAME OF YOUR SCRIPT GOES HERE }
```

**Reply by Wolfang42** [reply]
Could it work the same way if I made a button in Node-Red to call the Alexa routine?

**Reply by PRProd** [reply]
I'm unfamiliar with Node-Red, so I can't answer that for you.

**Reply by Wolfang42** [reply]
It is a way to build automations in HA visually rather than with coding. I don't know how to code in HA very well at all to make the script necessary.  But I know how to work an automation flow in Node-Red within home assistant to get things to work.

**Reply by Wolfang42** [reply]
ok with the help of ChatGPT and some mistakes, I have it working!  Is there a way to have the power button toggle off as well as on?  I have 2 routines, one for off and one for on.

**Reply by PRProd** [reply]
If you feel comfortable posting your "on" and "off" scripts here, I can make something for you to try.

**Reply by Wolfang42** [reply]
```
den_roku_on:
  alias: "Den Roku On"
  sequence:
    - service: media_player.play_media
      target:
        entity_id: media_player.kitchen_echo # Replace with your Alexa device's entity ID
      data:
        media_content_id: "Den Roku" # Replace with the exact name of your Alexa routine
        media_content_type: "routine"

den_roku_off:
  alias: "Den Roku Off"
  sequence:
    - service: media_player.play_media
      target:
        entity_id: media_player.kitchen_echo # Replace with your Alexa device's entity ID
      data:
        media_content_id: "turn off den" # Replace with the exact name of your Alexa routine
        media_content_type: "routine"

**Reply by PRProd** [reply]
@Wolfang42 ~

I don't have any way of testing this on my end, so I can't guarantee anything.  Here goes attempt 1.  :smile: 

After replacing the appropriate entity_id where you see {{ THE ENTITY NAME OF YOUR PANASONIC TV GOES HERE }}, create and save this script.
```yaml
alias: "Toggle Den Roku Power"
sequence:
  - if:
      - condition: state
        entity_id: {{ THE ENTITY NAME OF YOUR PANASONIC TV GOES HERE }}
        state: "on"
    then:
      - service: media_player.play_media
        target:
          entity_id: media_player.kitchen_echo
        data:
          media_content_id: "turn off den"
          media_content_type: "routine"
    else:
      - service: media_player.play_media
        target:
          entity_id: media_player.kitchen_echo
        data:
          media_content_id: "Den Roku"
          media_content_type: "routine"
description: ""
icon: mdi:power
```

IF by some miracle running that script toggles the power on and off, then simply add this to your Firemote YAML:

```yaml
button_overrides:
  power-button:
    script: toggle_den_roku_power
```

Of course, if it doesn't work - I'm persistent.  Just let me know.

**Reply by PRProd** [reply]
The suspense is killing me... I really want to score a win for you before I'm away for two weeks.

**Reply by Wolfang42** [reply]
Sorry I was gone for a couple of days.  Your script works perfectly!  There is a problem that the Panasonic Viera integration will not keep the state of the TV updated at all times, and so when it is updated, and the TV is on, the script turns it off perfectly. But when the integration loses the state of the TV (it goes to off when it isn't) the script obviously doesn't realize the TV is on still.  I'm not sure how to fix that part. Thank you so much for all this!

**Reply by PRProd** [reply, marked as answer]
AWESOME!!!!  :tada: :partying_face: 

Now you have options!  

1. Keep this one & deal with the silly integration losing access
```yaml
button_overrides:
  power-button:
    script: toggle_den_roku_power
```
2. Keep this one and add a new button that can be used to turn it off
```yaml
custom_launchers:
  - friendly_name: power off
    label: 'Off'
    color: white
    background: '#662d91'
    script: den_roku_off
app_launch_1: customlauncher power off
``` 
3. Rewire your network tomorrow.  :smile: 

Seriously though, I'm glad this is working now!  I didn't doubt it for a second.

**Reply by Wolfang42** [reply]
Aww I need help rewiring the network!  So many CAT6 wires everywhere and not labeled by the Savant integrator.  They were awful.

Where does the second script go?

**Reply by PRProd** [reply]
The second block of YAML goes in your Firemote YAML config if you want to use it.

You can replace the "app_launch_1" part with wherever you'd like to place it instead.

https://a.co/d/5IkVr14

**Reply by Wolfang42** [reply]
OMG that works great!!!  I had to move it to 6th slot. It's perfect!  I cannot thank you enough for teaching me so much. This will all be best once Savant is extracted and I can create new scenes.

Oh, I'll look at that amazon purchase if I ever get the courage to dive into the mess. 

Up for another challenge? I don't want to wear out my welcome but just in case you have ideas after your trip:

Those 1 and 2 buttons look perfect for programming volume.  The volume for this TV is an LG sound bar.  That is currently controlled via IR blaster upstairs with the Savant system, that is somehow transmitted locally and I see the emitter in front of the soundbar.  Is it possible to use these 2 buttons to program the volume control? I know this LG soundbar also has BT and WiFi connectivity as well but right now receives audio from an optical line out of the TV. Or is the solution to get a different IR blaster like your first post mentioned?

**Reply by PRProd** [reply]
First off, YAY!!!  Super glad that the pesky power situation is figured out now!!

The Amazon thing - that was kind of a joke, but kind of not?  A device like that is quite useful when trying to figure out which unlabeled wire is which... assuming you have an excellent label maker with you when you're doing it.

So, onto the next puzzle.  😄  (someone like you with patience and an eagerness to learn will never wear out their welcome)
<br>

### The Magic Formula
If you can control a thing with Home Assistant somehow, then we can make a Firemote button tell Home Assistant to do that thing.

If that answer right now is "no, there's not an integration that can control the soundbar" then there still might be a different route.  If your TV has an [ARC or eARC](https://www.howtogeek.com/763334/what-is-earc/) HDMI port, and your soundbar has a matching HDMI/ARC input port, then perhaps the volume can be controlled from Home Assistant via the Panasonic integration.

I'm still around for the workday tomorrow, but then I'll be gone for almost two weeks.  I'll be able to respond to your questions for at least the next 20 hours or so.  👍🏻

**Reply by Wolfang42** [reply]
Well, this is cool! I looked at the soundbar (LG SH7B, now discontinued) and it has BT, WiFi,  HDMI and Ethernet connectivity.  I'm sure I could try the HDMI ARC but since the Panasonic integration keeps losing connection, I figured that wouldn't likely work. I cannot connect the WiFi of the soundbar since the LG software does not support this device any longer. So on a whim I connected an ethernet cable to it and wham! Home Assistant and Alexa both notified me of its presence.  So maybe it will work with the 1-2 buttons on the Firemote to control its volume?

**Reply by Wolfang42** [reply]
Have a great 2 weeks off!

**Reply by PRProd** [reply]
@Wolfang42 ~

I'm back!  😄  How did it go?  Is everything "Firemote-ed" in your world now?

**Reply by Wolfang42** [reply]
Welcome back! There is indeed the HDMI connection I can use between the TV and the soundbar, but the HA Panasonic integration is too flaky and loses connection to the TV so that option won't work. I think an IR blaster is the only solution perhaps unfortunately. HA recognizes the soundbar but as a device to cast to, not volume control :(

**Reply by PRProd** [reply]
Oh man!  That stinks!  If that would have been known in the beginning, you could have just started there and it would have answered all of the things.  Sorry to hear that, truly I am.

**Reply by Wolfang42** [reply]
No way! This has been a great success...I have the Firemote turning on and off the system. It's glorious! The volume is an easy matter with the IR blaster I assume? Currently that is how the Savant system is controlling it, but the blaster is upstairs and routed to the emitter at the soundbar somehow.

**Reply by PRProd** [reply]
I've only dipped my toe in the waters of IR blasters after #405 was opened a couple of months ago.  The setup for a Broadlink device is cumbersome, but after it's configured to work with Home Assistant, it seems to be rock solid.

Are you saying that you have a completely separate IR blaster that isn't a native part of the Savant ecosystem?  If so, I don't suppose you've been able to get HA to find that, have you?

**Reply by Wolfang42** [reply]
I think it is a Savant device but I could be wrong. I'll  trudge upstairs and try to locate it and take a pic.
